src/main/java/org/apache/commons/text/similarity/SorensenDiceSimilarity.java

```diff
@@ -0,0 +1,127 @@
+ * similarity of two samples. It was independently developed by the botanists
+ * Thorvald S&#248;rensen and Lee Raymond Dice, who published in 1948 and 1945
+ * respectively. The index is known by several other names, especially
+ * S&#248;rensen–Dice index, S&#248;rensen index and Dice's coefficient. Other
+ * variations include the "similarity coefficient" or "index", such as Dice
+ * similarity coefficient (DSC).
+ * </p>
+ *
+ * <p>
+ * This implementation is based on the S&#248;rensen–Dice similarity algorithm
+ * from <a href=
+ * "http://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient">
+ * http://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient</a>.
+ *
+ *
+ * </p>
+ *
+ * @since 1.7
+ */
+public class SorensenDiceSimilarity implements SimilarityScore<Double> {
+
+    /**
+     * Calculates Sorensen-Dice Similarity of two character sequences passed as
+     * input.
+     *
+     * <pre>
+     * similarity.apply(null, null)                 = IllegalArgumentException
+     * similarity.apply("foo", null)                = IllegalArgumentException
+     * similarity.apply(null, "foo")                = IllegalArgumentException
+     * similarity.apply("night", "nacht")           = 0.25
+     * similarity.apply("", "")                     = 1.0
+     * similarity.apply("foo", "foo")               = 1.0
+     * similarity.apply("foo", "foo ")              = 0.8
+     * similarity.apply("foo", "foo ")              = 0.66
+     * similarity.apply("foo", " foo ")             = 0.66
+     * similarity.apply("foo", " foo")              = 0.66
+     * similarity.apply("", "a")                    = 0.0
+     * similarity.apply("aaapppp", "")              = 0.0
+     * similarity.apply("frog", "fog")              = 0.4
+     * similarity.apply("fly", "ant")               = 0.0
+     * similarity.apply("elephant", "hippo")        = 0.0
+     * similarity.apply("hippo", "elephant")        = 0.0
+     * similarity.apply("hippo", "zzzzzzzz")        = 0.0
+     * similarity.apply("hello", "hallo")           = 0.5
+     * similarity.apply("ABC Corporation", "ABC Corp") = 0.7
+     * similarity.apply("D N H Enterprises Inc", "D &amp; H Enterprises, Inc.") = 0.74
+     * similarity.apply("My Gym Children's Fitness Center", "My Gym. Childrens Fitness") = 0.81
+     * similarity.apply("PENNSYLVANIA", "PENNCISYLVNIA") = 0.69
+     * </pre>
+     *
+     * @param left  the first CharSequence, must not be null
+     * @param right the second CharSequence, must not be null
+     * @return result similarity
+     * @throws IllegalArgumentException if either CharSequence input is {@code null}
+     */
+    @Override
+    public Double apply(final CharSequence left, final CharSequence right) {
+
+        if (left == null || right == null) {
+            throw new IllegalArgumentException("CharSequences must not be null");
+        }
+
+        if (StringUtils.equals(left, right)) {
```

Comment:
The `JaccardSimilarity` computes 0 for the equivalent input. 

This check should be moved after the check of the lengths. Thus any input with no bigrams gets a similarity of 0 since there is effectively nothing to compare.

Reply:
+1, this would be a more homogeneous approach for the library. We can review this later for 2.0 if necessary.

Reply:
@kinow I that case ```JaccardSimilarity``` must be fixed because nothing is always equal to nothing and is 100% similar to each other, right ?

also if I move length check above to equal check below test case fails, in below case we clearly see ```a=a``` and it is 100% similar.
```
        assertEquals(1d, similarity.apply("a", "a"));
``` 

I checked few other libraries to confirm.
```
amey@xps:~/work/repo/commons-text$ python
Python 2.7.15rc1 (default, Apr 15 2018, 21:51:34) 
[GCC 7.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from difflib import SequenceMatcher
>>> SequenceMatcher(None, "a", "a").ratio()
1.0
>>> 
>>> SequenceMatcher(None, "", "").ratio()
1.0
```

Reply:
```
>>> 1 - jellyfish.levenshtein_distance(u'a', u'a')
1
>>> 
>>> 1 - jellyfish.levenshtein_distance(u'', u'')
1
```

_Note -  similarity  = 1 - distance_

Reply:
I think we have a bug in `Jaccard` for empty strings @ameyjadiye . Still need to confirm with the algorithm definition, just to make sure. But the following code:

```java
package br.eti.kinoshita.tests.text;

import java.util.Collections;

public class EditDistances {

    public static void main(String[] args) {
        System.out.println("Testing jaccard sim/dis with empty strings");
        System.out.println("---");
        org.simmetrics.metrics.Jaccard<String> j1 = new org.simmetrics.metrics.Jaccard<>();
        float s1 = j1.compare(Collections.emptySet(), Collections.emptySet());
        System.out.println("Simmetrics Jaccard similarity: " + s1);
        float d1 = j1.distance(Collections.emptySet(), Collections.emptySet());
        System.out.println("Simmetrics Jaccard distance: " + d1);
        
        System.out.println("---");
        
        info.debatty.java.stringsimilarity.Jaccard j2 = new info.debatty.java.stringsimilarity.Jaccard();
        double s2 = j2.similarity("", "");
        System.out.println("javastringsimilarity Jaccard similarity: " + s2);
        double d2 = j2.distance("", "");
        System.out.println("javastringsimilarity Jaccard distance: " + d2);
        
        System.out.println("---");
        
        org.apache.commons.text.similarity.JaccardSimilarity j3_1 = new org.apache.commons.text.similarity.JaccardSimilarity();
        double s3 = j3_1.apply("", "");
        System.out.println("commons-text Jaccard similarity: " + s3);
        org.apache.commons.text.similarity.JaccardDistance j3_2 = new org.apache.commons.text.similarity.JaccardDistance();
        double d3 = j3_2.apply("", "");
        System.out.println("commons-text Jaccard distance: " + d3);
    }
}
```

Produces:

```
Testing jaccard sim/dis with empty strings
---
Simmetrics Jaccard similarity: 1.0
Simmetrics Jaccard distance: 0.0
---
javastringsimilarity Jaccard similarity: 1.0
javastringsimilarity Jaccard distance: 0.0
---
commons-text Jaccard similarity: 0.0
commons-text Jaccard distance: 1.0
```

>I checked few other libraries to confirm.

Did you compare Jaccard and Sorensen (the two discussed here) too?

>Note - similarity = 1 - distance

I think you meant `distance = 1 - similarity` @ameyjadiye , (as in [here](https://github.com/apache/commons-text/blob/ea15c5a76eba199f3c08bab96ea17981c0ec01c1/src/main/java/org/apache/commons/text/similarity/JaccardDistance.java#L53)). But you are again correct here :+1: 

>This check should be moved after the check of the lengths. Thus any input with no bigrams gets a similarity of 0 since there is effectively nothing to compare.

@aherbert, the Python `textdistance` is also returning 1 for both single chars and bigrams.

```python
>>> textdistance.sorensen("", "")
1
>>> textdistance.Sorensen(qval=2)("", "")
1
```

Same for java-string-similarity's `SorensenDice#similarity`. So I think @ameyjadiye 's code is actually OK for the current approach?

Reply:
Created a ticket to check the case for empty strings using Jaccard: https://issues.apache.org/jira/browse/TEXT-158

Reply:
So I reckon we don't have to do anything more and can resolve this conversation @ameyjadiye , @aherbert ?